text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+
+- Using UTF-8 instead of Punycode would make mangled strings containing non-ASCII identifiers a bit more human-readable. For demangled strings, there would be no difference.
+
+- Punycode support is non-optional since some platforms only allow a very limited character set for symbol names. Thus, we would be using UTF-8 on some platforms and Punycode on others, making it harder to predict what a symbol name for a given item looks like.
+
+- Punycode encoding and decoding is more runtime effort for the mangler and demangler.
+
+- Once a demangler supports Punycode, it is not much effort to support both encodings. The `u` identifier suffix tells the demangler whether it's Punycode. Otherwise it can just assume UTF-8 which already subsumes ASCII.
+
+### Re-use <disambiguator> for crate disambiguator
+
+The RFC currently proposes to represent crate-ids as an `<identifier>` of the form `<name>_<disambiguator>`. However, the `<identifier>` production already supports disambiguation via its `<disambiguator>` component. The crate disambiguator could be encoded into an disambiguation index.
+
+# Appendix A - Suggested Demangling
+
+This RFC suggests that names are demangled to a form that matches Rust syntax as it is used in source code, compiler error messages and `rustdoc`:
+
+ - Path components should be separated by `::`.
+
+ - If the path root is a `<crate-id>` it should be printed as the crate name. If the context requires it for correctness, the crate disambiguator should be printed too, as in, for example, `std[a0b1c2d3]::collections::HashMap`. In this case `a0b1c2d3` would be the disambiguator. Usually, the disambiguator can be omitted for better readability.
+
+ - If the path root is a trait impl, it should be printed as `<SelfType as Trait>`, like the compiler does in error messages.
+
+ - The list of generic arguments should be demangled as `<T1, T2, T3>`.
+
+ - Identifiers and trait impl path roots can have a numeric disambiguator (the `<disambiguator>` production). The syntactic version of the numeric disambiguator maps to a numeric index. If the disambiguator is not present, this index is 0. If it is of the form `s_` then the index is 1. If it is of the form `s<base-62-digit>_` then the index is `<base-62-digit> + 2`. The suggested demangling of a disambiguator is `[<index>]`. However, for better readability, these disambiguators should usually be omitted in the demangling altogether. Disambiguators with index zero can always be omitted.
+
+ The exception here are closures. Since these do not have a name, the disambiguator is the only thing identifying them. The suggested demangling for closures is thus `{closure}[<index>]`.
```

Comment:
Similarly here, we should avoid braces. What does C++ do for its lambdas?

Reply:
GCC uses something with braces and indices too:
```cpp
int square(int num) {
    auto foo = [num]() -> int { return num * num; };
    return foo();
}
```
The closure is demangled as `square(int)::{lambda()#1}::operator()() const`
(see https://godbolt.org/z/TaXWCe)

Reply:
Do debuggers work well with it? If so, how? Can we do some tests to see what works and what doesn't?

Reply:
I assume that debuggers treat lambdas as regular `operator()` methods. What kind of tests did you have in mind?

Reply:
I'm referring to the problems @m4b mentions in https://github.com/rust-lang/rfcs/pull/2603#discussion_r236779961, regarding debuggers not being able to let you refer to symbol names that contain `{` (or perhaps only `{{`?).

If we change `{{closure}}` in the compiler with some other notation, we can see how well GDB and LLDB interact with the symbol names.

Although it's possible debuggers only handle such symbol names when they come from a mangling, which would mean debuggers should just pick a demangling that works for them, right?